File: tests/claim-extractor.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { createClaimExtractor } from "../src/infrastructure/claim-extractor.js";

const source = {
  text: "She wrote the novel.",
  title: "Mary Shelley",
  context: { sectionPath: ["Frankenstein"], leadIn: "Shelley began writing in 1816." },
};
const extractor = (program: string) =>
  createClaimExtractor(process.execPath, "-e", program, "unused", "unused", 5000);

test("an extractor that exits before reading a large source rejects without crashing the host", async () => {
  await assert.rejects(
    extractor("process.exit(1)")(
      { ...source, text: "x".repeat(900_000) },
      AbortSignal.timeout(5000),
    ),
    /EPIPE|ECONNRESET|ClaimExtractor failed/u,
  );
});

test("the extractor passes unchanged text and bounded context to its worker", async () => {
  const claims = await extractor(`
    let input = '';
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', chunk => input += chunk);
    process.stdin.on('end', () => {
      process.stdout.write(JSON.stringify({ claims: [{ content: input }] }));
    });
  `)(source, AbortSignal.timeout(5000));
  assert.deepEqual(JSON.parse(claims[0]!), source);
});

File: tests/claim_extractor_test.py
import io
import json
import runpy
import unittest
from pathlib import Path
from unittest.mock import patch


class ClaimExtractorContextTest(unittest.TestCase):
    def test_context_resolves_references_without_replacing_selected_text(self):
        source = {
            "text": "  She wrote the novel.\n",
            "title": "Mary Shelley",
            "context": {
                "sectionPath": ["Frankenstein"],
                "leadIn": "Shelley began writing in 1816.",
            },
        }
        extraction = {"extractions": {"claims": [
            {"content": "Mary Shelley wrote Frankenstein.", "subtype": "Factoid"},
        ]}}
        response = io.BytesIO(json.dumps({"choices": [{
            "message": {"content": json.dumps(extraction)},
        }]}).encode())
        with (
            patch("sys.argv", ["claim_extractor.py", "http://localhost/v1", "test", "UNUSED_KEY", "5"]),
            patch("sys.stdin", io.StringIO(json.dumps(source))),
            patch("sys.stdout", new_callable=io.StringIO) as output,
            patch("urllib.request.urlopen", return_value=response) as request,
        ):
            runpy.run_path(str(Path(__file__).parents[1] / "python/claim_extractor.py"), run_name="__main__")
        body = json.loads(request.call_args.args[0].data)
        system, user = body["messages"]
        self.assertIn("Do not extract additional claims from the context", system["content"])
        self.assertIn(json.dumps({"title": source["title"], "context": source["context"]}), user["content"])
        self.assertIn("LAST MESSAGE (ASSISTANT):\n" + source["text"], user["content"])
        self.assertEqual(json.loads(output.getvalue()), {"claims": [{"content": "Mary Shelley wrote Frankenstein."}]})


if __name__ == "__main__":
    unittest.main()
